src/Badges.sol

```diff
@@ -291,23 +321,46 @@ contract Badges is
     return tokenId;
   }
 
-  // signature comes from the people with the right params
+  // todo:: do we need to use a sparse markle tree? Is it more gas efficient?
+  // todo:: handling vouchers - what would be a voucherID here?
+  function safeCheckMerkleAgreement(
+    address _to,
+    address _from,
+    string calldata _uri,
+    bytes calldata _signature,
+    bytes32 _root,
+    bytes32[] calldata _proof
+  ) public view virtual returns (uint256) {
+    // this authenticates the signature coming from the issuer
+    bytes32 hash = getMerkleAgreementHash(_from, _uri, _root);
+    require(
+      SignatureCheckerUpgradeable.isValidSignatureNow(_from, hash, _signature),
+      "safeCheckMerkleAgreement: invalid signature"
+    );
+    bytes32 leaf = keccak256(abi.encodePacked(_to));
+    // this authenticates that the claimant is indeed part of the tree whose root was signed
+    require(MerkleProof.verify(_proof, _root, leaf), "safeCheckMerkleAgreement: invalid leaf");
+    uint256 voucherHashId = uint256(keccak256(abi.encode(_from, _to)));
```

Comment:
We're not using 2 different ways to generate a voucherHashId... 

1. would there be collissions you think?
2. if you get a badge via multi-issuance, would the protocol prevent the user from having a badge via the single-issuance?

Reply:
1. I don't think there would be collisions... but to maximize probability of uniqueness, we might as well just throw in the `MERKLE_AGREEMENT_HASH` and `uri`... wouldn't hurt. 
2. with that, we definitely won't have collisions or issues with single/multi-issuance conflicting